### PR TITLE
Added site cache warming and cache switching.

### PR DESCRIPTION
**Pull request by avasquez614**
- Added site cache warming and cache switching.
- Enhancements to Context creation and initialization.
- Enhancements to the DeploymentEventsWatcher.
- Other general enhancements.

Ticket craftercms/craftercms#3472

Depends on craftercms/core#112

### CODE REVIEW

File: src/main/resources/crafter/engine/mode/preview/services-context.xml [hunk 1]
```diff
@@ -69,15 +69,17 @@
         <property name="groovyClassesPath" value="${crafter.engine.site.default.classes.groovy.path}"/>
         <property name="cacheOn" value="false"/>
         <property name="urlTransformationEngine" ref="crafter.urlTransformationEngine"/>
-        <property name="overlayCallback" ref="crafter.authoringOverlayCallback"/>
```
Comment (jdrossl) [inline src/main/resources/crafter/engine/mode/preview/services-context.xml]: I think we need to update the Freemarker templates because of this right?

Comment (avasquez614) [reply src/main/resources/crafter/engine/mode/preview/services-context.xml]: I didn't see any overlaycallback in the templates

Comment (avasquez614) [reply src/main/resources/crafter/engine/mode/preview/services-context.xml]: Darn yes there is one ok let me check

Comment (avasquez614) [reply src/main/resources/crafter/engine/mode/preview/services-context.xml]: It's only used in the FTLs to check if the Engine is running in preview mode, I'll create a ticket to fix that.

**Pull request merged**